beluga/include/beluga/algorithm/distance_map.hpp

```diff
@@ -92,6 +92,53 @@ auto nearest_obstacle_distance_map(
   return distance_map;
 }
 
+/// A distance map function considering unknown values.
+template <class Range, class DistanceFunction, class NeighborsFunction>
+auto nearest_obstacle_unknown_distance_map(
+    Range&& obstacle_map,
+    DistanceFunction&& distance_function,
+    NeighborsFunction&& neighbors_function,
+    const double unknown_distance_map) {
+  struct IndexPair {
+    std::size_t nearest_obstacle_index;
+    std::size_t index;
+  };
+
+  using DistanceType = std::invoke_result_t<DistanceFunction, std::size_t, std::size_t>;
+  auto distance_map = std::vector<DistanceType>(ranges::size(obstacle_map));
+  auto visited = std::vector<bool>(ranges::size(obstacle_map), false);
+
+  auto compare = [&distance_map](const IndexPair& first, const IndexPair& second) {
+    return distance_map[first.index] > distance_map[second.index];
+  };
+  auto queue = std::priority_queue<IndexPair, std::vector<IndexPair>, decltype(compare)>{compare};
+
+  for (auto [index, tuple_data] : ranges::views::enumerate(obstacle_map)) {
+    if (std::get<0>(tuple_data)) {
+      visited[index] = true;
+      distance_map[index] = 0;
+      queue.push(IndexPair{index, index});  // The nearest obstacle is itself
+    } else if (std::get<1>(tuple_data)) {
+      visited[index] = true;
```

Comment:
@DPR00 hmm, won't this partially (or totally) block the greedy expansion below? Thinking about this again, it may be safer to override the distance map at the indices we know are actually unknown after computing it in full. We are doing this once, so the performance penalty shouldn't be that bad.